Activity mgmt

aaronmsegal opened the pull request:
Adds the ability for a Teacher to add activities to the classes they are teaching.

Changes:
1. Added getBook endpoint.
2. Added getBookList endpoint, which returns a list of all bookId : bookTitle pairs.
3. Added getSampleProblems endpoint, which returns up to 4 sample problems, generated from the templates in the given book, chapter, and section.
4. Added addActivity endpoint which adds an activity to the given class, if the user is the teacher of the class.
5. Added book_ctrl.getBookList for getBookList endpoint.
6. Added class_ctrl.addActivity  for addActivity endpoint.
7. Modified ResponseCreation.py to allow for inclusion of decimals in json dump
8. Changed how Activities are displayed 
9. Added getBook, getAllBookTitlesAndIds getSampleProblems to book.service for above endpoints.
10. Added addActivity to class.service for above endpoint.
10. Created the Add Activity modal for the class details page.

How To Test:
1. Login as admin
2. Create a book with at least one chapter, section, and problem
3. Create a class, go to the details page for that class
4. Click the add new activity button
5. Confirm the add activity modal appears and functions as expected
6. Select your new book from the book drop down, and then the desired chapter and section
7. Confirm that 1-4 problems display from the section
8. Click the add new activity button on the modal
9. Confirm the new activity has been added to the class
10. repeat steps 4-9, but this time use the chapter 1 and section 1 of the Foundations of Algebra I text book
11. Try to mentor mentii from both of the activities you have created.

jonathanmcdaniel commented:
Two things I've noticed while working through this. Both are pretty small. 
1) There is an extra space between how many problems to do and the due date
"Try to mentor mentii through 5 problems , before 2017-05-15"

2) Could we add a spinner to the sample problem section when loading from a new section? Also maybe immediately hide the sample problems from the old section until the new one has loaded?

smc395 commented:
Three issues:
1. If I create an activity with a start date, but no due date, a default due date of the same start date is made and shown (see Micah's Class)
2. If I set a due date with no start date, the due date doesn't show up in the panel
3. When I select a start date and a different due date, the due date is set to the start date, not the date I specified.

smc395 commented:
How hard would it be to add functionality to delete activities?

smc395 commented on Backend/app.py:
```diff
@@ -306,6 +306,84 @@ def createBook():
+@auth.login_required
+@handleOptionsRequest
+def getSampleProblems(bookId, chapterTitle, sectionTitle):
+  print bookId, chapterTitle, sectionTitle
```
remove in final code

smc395 commented on Backend/utils/ResponseCreation.py:
```diff
@@ -44,6 +48,8 @@ def prepForJsonDump(self, item):
       for i in iterator:
         #recursive call
         item[i] = self.prepForJsonDump(item[i])
+    elif isinstance(item, decimal.Decimal): #also numbers cant be dumped ??!
```
what does this do?

aaronmsegal replied:
Apparently json.dumps cannot handle the number for problem count (or anything else) unless it is first cast to a string

smc395 commented on Backend/utils/ResponseCreation.py:
```diff
@@ -26,6 +26,10 @@ def addToPayload(self, attribute, value):
     value = self.prepForJsonDump(value)
     self.payload[attribute] = value
 
+  def setPayload(self, payload):
```
what is this for?

smc395 commented on Frontend/app/class/detail/detail.component.ts:
```diff
@@ -66,11 +84,153 @@ export class ClassDetailComponent implements OnInit, OnDestroy {
 
+  /* Activity Modal Methods */
+  showActivityModal():void {
+    this.newActivity = new ActivityModel('', undefined, 5, undefined, undefined, undefined, undefined, undefined);
```
is there a reason you are using undefined instead of just ` ' ' ` for strings in the model?

aaronmsegal replied:
making it undefined makes it the default item in the select.

smc395 commented on Frontend/app/class/detail/detail.component.ts:
```diff
@@ -66,11 +84,153 @@ export class ClassDetailComponent implements OnInit, OnDestroy {
+    this.sampleProblems = [];
+  }
+
+  booksRecived(books) {
```
spelling: booksReceived

smc395 commented on Frontend/app/class/detail/detail.component.ts:
```diff
@@ -66,11 +84,153 @@ export class ClassDetailComponent implements OnInit, OnDestroy {
+      );
+  }
+
+  bookRecived(book) {
```
spelling: bookReceived

smc395 commented on Frontend/app/class/detail/detail.component.ts:
```diff
@@ -66,11 +84,153 @@ export class ClassDetailComponent implements OnInit, OnDestroy {
+        this.newActivity.chapterTitle,
+        this.newActivity.sectionTitle)
+      .subscribe(
+        data => this.problemsRecived(data.json().payload.problems),
```
spelling: problemsReceived

aaronmsegal commented:
@jonathanmcdaniel I fixed the spacing, the spinner and removing the sample problems when switching sections

@smc395 I fixed issues with due date and start date display. The ability to delete activities is complex enough as to require another story..
